be/src/exec/pipeline/exchange/mem_limited_chunk_queue.cpp

```diff
@@ -0,0 +1,539 @@
+    if (unconsumed_bytes >= _opts.max_unconsumed_bytes) {
+        return false;
+    }
+
+    size_t in_memory_bytes = _total_accumulated_bytes - _flushed_accumulated_bytes;
+    if (in_memory_bytes >= _opts.memory_limit && _next_flush_block->next != nullptr) {
+        if (bool expected = false; _has_flush_io_task.compare_exchange_strong(expected, true)) {
+            TEST_SYNC_POINT("MemLimitedChunkQueue::can_push::before_submit_flush_task");
+            auto status = _submit_flush_task();
+            if (!status.ok()) {
+                _update_io_task_status(status);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    return true;
+}
+
+StatusOr<ChunkPtr> MemLimitedChunkQueue::pop(int32_t consumer_index) {
+    TEST_SYNC_POINT("MemLimitedChunkQueue::pop::before_pop");
+    DCHECK(consumer_index <= _consumer_number);
+    RETURN_IF_ERROR(_get_io_task_status());
+    std::unique_lock l(_mutex);
+
+    DCHECK(_consumer_progress[consumer_index] != nullptr);
+    auto iter = _consumer_progress[consumer_index].get();
+    if (!iter->has_next()) {
+        if (_opened_sink_number == 0) {
+            return Status::EndOfFile("no more data");
+        }
+        return Status::InternalError("unreachable path");
+    }
+    iter->move_to_next();
+    DCHECK(iter->get_block()->in_mem);
+    Cell* cell = iter->get_cell();
+    DCHECK(cell->chunk != nullptr);
+
+    cell->used_count += 1;
+    VLOG_ROW << fmt::format(
+            "[MemLimitedChunkQueue] pop chunk for consumer[{}] from block[{}], accumulated_rows[{}], "
+            "accumulated_bytes[{}], used_count[{}]",
+            consumer_index, (void*)(iter->get_block()), cell->accumulated_rows, cell->accumulated_bytes,
+            cell->used_count);
+    iter->get_block()->pending_read_requests--;
+    DCHECK(iter->get_block()->pending_read_requests >= 0);
+    auto result = cell->chunk;
+
+    _update_progress(iter);
+    return result;
+}
+
+void MemLimitedChunkQueue::_evict_loaded_block() {
+    if (_loaded_blocks.size() < _consumer_number) {
+        return;
+    }
+    Block* block = _loaded_blocks.front();
+    _loaded_blocks.pop();
+    _current_load_rows -= block->flush_rows;
+    _current_load_bytes -= block->flush_bytes;
+    for (auto& cell : block->cells) {
+        if (cell.chunk != nullptr) {
+            cell.chunk.reset();
+        }
+    }
+    block->in_mem = false;
+    VLOG_ROW << fmt::format("[MemLimitedChunkQueue] evict block [{}]", (void*)block);
+}
+
+bool MemLimitedChunkQueue::can_pop(int32_t consumer_index) {
+    DCHECK(consumer_index < _consumer_number);
+    std::shared_lock l(_mutex);
+    DCHECK(_consumer_progress[consumer_index] != nullptr);
+
+    auto iter = _consumer_progress[consumer_index].get();
+    if (iter->has_next()) {
+        auto next_iter = iter->next();
+        if (next_iter.block->in_mem) {
+            next_iter.block->pending_read_requests++;
```

Comment:
what will happen if next_iter.block called in pipeline poller?

Reply:
> what will happen if can_pop called in pipeline poller?

fixed by using BitMask to ensure idempotency